src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolWorkQueue.cs

```diff
@@ -409,9 +424,126 @@ private struct CacheLineSeparated
 
         public ThreadPoolWorkQueue()
         {
+            for (int i = 0; i < AssignableWorkItemQueueCount; i++)
+            {
+                assignableWorkItemQueues[i] = new ConcurrentQueue<object>();
+            }
+
             RefreshLoggingEnabled();
         }
 
+        private void AssignWorkItemQueue(ThreadPoolWorkQueueThreadLocals tl)
+        {
+            Debug.Assert(AssignableWorkItemQueueCount > 0);
+
+            _queueAssignmentLock.Acquire();
+
+            // Determine the first queue that has not yet been assigned to the limit of worker threads
+            int queueIndex = -1;
+            int minCount = int.MaxValue;
+            int minCountQueueIndex = 0;
+            for (int i = 0; i < AssignableWorkItemQueueCount; i++)
+            {
+                int count = _assignedWorkItemQueueThreadCounts[i];
+                Debug.Assert(count >= 0);
+                if (count < ProcessorsPerAssignableWorkItemQueue)
+                {
+                    queueIndex = i;
+                    _assignedWorkItemQueueThreadCounts[queueIndex] = count + 1;
+                    break;
+                }
+
+                if (count < minCount)
+                {
+                    minCount = count;
+                    minCountQueueIndex = i;
+                }
+            }
+
+            if (queueIndex < 0)
+            {
+                // All queues have been fully assigned. Choose the queue that has been assigned to the least number of worker
+                // threads.
+                queueIndex = minCountQueueIndex;
+                _assignedWorkItemQueueThreadCounts[queueIndex]++;
+            }
```

Comment:
It seems like this is trying to pack threads into queues, preferring to fill up a queue before assigning to another one?  I'm curious about that approach rather than trying to just pick whichever queue has the fewest threads assigned to it?

Reply:
Always picking the queue with the fewest threads assigned seemed like it could be problematic, so I avoided it. For instance there may be a large batch of IO events queued as one work item in the global queue. One thread picks it up and parallelizes processing of that batch in its assigned queue. Threads assigned to different queues would be forced to search other assigned queues to find any work. Or there may be new smaller batches of IO events later in the global queue, then threads process those instead of helping with the first larger batch, which can lead to a bit of out-of-order processing. I wanted to keep it simple for now. It may be feasible to always assign the least-assigned queue if we handle IO events a bit differently, could experiment with that separately.